### PR TITLE
cgroup: add support for memory.peak

### LINKED ISSUE #29878: Add support for cgroup's memory.peak value
### Component

systemctl, systemd

### Is your feature request related to a problem? Please describe

Systemd should expose the `memory.peak` value of cgroups with memory accounting.

```
● NetworkManager.service - Network Manager
     Loaded: loaded (/usr/lib/systemd/system/NetworkManager.service; enabled; preset: disabled)
     Active: active (running) since Mon 2023-11-06 10:15:54 CET; 3h 8min ago
       Docs: man:NetworkManager(8)
   Main PID: 736 (NetworkManager)
      Tasks: 4 (limit: 38058)
     Memory: 15.9M (peak: 16.4M)
        CPU: 375ms
     CGroup: /system.slice/NetworkManager.service
             └─736 /usr/sbin/NetworkManager --no-daemon
```

### Describe the solution you'd like

_No response_

### Describe alternatives you've considered

_No response_

### The systemd version you checked that didn't have the feature you are asking for

_No response_

### PR DESCRIPTION
**Pull request by Flowdalic**
Linux's Control Group v2 interfaces exposes memory.peak, which contains the "max memory usage recorded for the cgroup and its descendants since the creation of the cgroup."

This commit adds a new property "MemoryPeak" for units and makes "systemctl show" display this value if it is available.

Fixes #29878.

### CODE REVIEW

**Comment by poettering** [comment]
looks good to me, but please run "ninja -C build update-dbus-docs" and fill in a word or to to the man page change this results in

**Comment by poettering** [comment]
I'd be fine with adding this to v255 still, seems like a fairly minor addition to me. But please add an entry for this to NEWS then.

**Comment by Flowdalic** [comment]
> but please run "ninja -C build update-dbus-docs"…

Done. Thanks for the hint!

> …and fill in a word or t[w]o to the man page change this results in

Besides the automatic change done by `update-dbus-docs`? To which man page should I add the word or two?

File: tools/dbus_ignorelist [hunk 1]
```diff
@@ -882,6 +882,7 @@ org.freedesktop.systemd1.Mount.ManagedOOMSwap
 org.freedesktop.systemd1.Mount.MemoryAccounting
 org.freedesktop.systemd1.Mount.MemoryAvailable
 org.freedesktop.systemd1.Mount.MemoryCurrent
+org.freedesktop.systemd1.Mount.MemoryPeak
```
Comment (YHNdnzj) [inline tools/dbus_ignorelist]: This file is for dbus methods / properties added before v250. For new properties please add a version info in the "History" section of org.freedesktop.systemd1.xml instead

**Review by YHNdnzj — CHANGES_REQUESTED** [review]

File: tools/dbus_ignorelist (hunk 1 above)
Comment (Flowdalic) [reply tools/dbus_ignorelist]: Done (hopefully correctly).

File: man/org.freedesktop.systemd1.xml [hunk 2]
```diff
@@ -11717,7 +11753,8 @@ $ gdbus introspect --system --dest org.freedesktop.systemd1 \
       <varname>ExtensionImagePolicy</varname> were added in version 254.</para>
       <para><varname>NFTSet</varname>,
       <varname>SetLoginEnvironment</varname> and
-      <varname>CoredumpReceive</varname> were added in version 255.</para>
+      <varname>CoredumpReceive</varname> and
+      <varname>MemoryPeak</varname> were added in version 255.</para>
```
Comment (YHNdnzj) [inline man/org.freedesktop.systemd1.xml]:
```
      <varname>SetLoginEnvironment</varname>,
      <varname>CoredumpReceive</varname>, and
      <varname>MemoryPeak</varname> were added in version 255.</para>
```

(please use Oxford comma and use "add" only for the last variable.)

**Review by YHNdnzj — APPROVED** [review]
One minor suggestion for style consistency

File: man/org.freedesktop.systemd1.xml (hunk 2 above)
Comment (Flowdalic) [reply man/org.freedesktop.systemd1.xml]: > please use Oxford comma

Done

> use "add" only for the last variable.

I am sorry, I do not understand what you meant?

**Comment by poettering** [comment]
Hmm, could you address #28542 at the same time? i.e. also fix up `unit_log_resources()` to include this information in the generated log message?

**Comment by Flowdalic** [comment]
> Hmm, could you address #28542 at the same time?

Yes, definitely. That was the use case I had in mind, but with Gentoo's `emerge`, not `dnf` ;-) 

However, I don't get it working. Not sure what is missing in https://github.com/systemd/systemd/pull/29879/commits/05c0c2402e243ea8f4926508d0682c92d599c9c6. Any pointers would be appreciated. Feels like I am missing something obvious.

**Comment by Flowdalic** [comment]
> Feels like I am missing something obvious.

Alright, I guess I need to cache the last memory peak, akin to `cpu_usage_last`.

**Comment by Flowdalic** [comment]
> Alright, I guess I need to cache the last memory peak, akin to `cpu_usage_last`.

Done

```bash
$ sudo systemd-run --wait -t --property=MemoryAccounting=1   sleep .1 
Running as unit: run-u246.service; invocation ID: d9a71ef36b844378b379f3ef49642389
Press ^] three times within 1s to disconnect TTY.
Finished with result: success
Main processes terminated with: code=exited/status=0
Service runtime: 115ms
CPU time consumed: 2ms
Memory peak: 256.0K
```

**Comment by poettering** [comment]
excellent! thanks!

**Review by YHNdnzj — APPROVED** [review]

**Pull request merged**